assemblers/org.osbuild.ostree.commit

```diff
@@ -0,0 +1,114 @@
+#!/usr/bin/python3
+
+import json
+import os
+import subprocess
+import sys
+import tempfile
+
+import osbuild.ostree
+
+
+STAGE_DESC = "Assemble a file system tree into a ostree commit"
+STAGE_INFO = """
+Takes a file system tree that is already conforming to the ostree
+system layout[1] and commits it to an archive repository.
+
+The repository is located at the `/repo` directory and additional
+metadata is stored in /commit.id and /compose.json which contain
+the commit id and the compose information respectively.
+
+[1] https://ostree.readthedocs.io/en/stable/manual/adapting-existing/
+"""
+
+TOPLEVEL_DIRS = ["dev", "proc", "run", "sys", "sysroot", "var"]
+TOPLEVEL_LINKS = {
+    "home": "var/home",
+    "media": "run/media",
+    "mnt": "var/mnt",
+    "opt": "var/opt",
+    "ostree": "sysroot/ostree",
+    "root": "var/roothome",
+    "srv": "var/srv",
+}
+
+
+def copy(name, source, dest):
+    print(f"cp {source}/{name} -> {dest}/{name}", file=sys.stderr)
+    subprocess.run(["cp", "--reflink=auto", "-a",
+                    os.path.join(source, name),
+                    os.path.join(dest, name)],
+                   check=True)
+
+
+def init_rootfs(root, tmp_is_dir):
+    """Initialize a pristine root file-system"""
+
+    fd = os.open(root, os.O_DIRECTORY)
+
+    os.fchmod(fd, 0o755)
+
+    for d in TOPLEVEL_DIRS:
+        os.mkdir(d, mode=0o755, dir_fd=fd)
+        os.chmod(d, mode=0o755, dir_fd=fd)
+
+    for l, t in TOPLEVEL_LINKS.items():
+        # <dir_fd>/l -> t
+        os.symlink(t, l, dir_fd=fd)
+
+    if tmp_is_dir:
+        os.mkdir("tmp", mode=0o1777, dir_fd=fd)
+        os.chmod("tmp", mode=0o1777, dir_fd=fd)
+    else:
+        os.symlink("tmp", "sysroot/tmp", dir_fd=fd)
+
+
+def main(tree, output_dir, options):
+    tmp_is_dir = options.get("tmp-is-dir", True)
+    parent = options.get("parent", None)
+
+    with tempfile.TemporaryDirectory(dir=output_dir) as root:
+        print("Initializing root filesystem", file=sys.stderr)
+        init_rootfs(root, tmp_is_dir)
+
+        print("Copying data", file=sys.stderr)
+        copy("usr", tree, root)
+        copy("boot", tree, root)
+        copy("var", tree, root)
+
+        for name in ["bin", "lib", "lib32", "lib64", "sbin"]:
+            if os.path.lexists(f"{tree}/{name}"):
+                copy(name, tree, root)
+
+        repo = os.path.join(output_dir, "repo")
+
+        subprocess.run(["ostree",
+                        "init",
+                        "--mode=archive",
+                        f"--repo={repo}"],
+                       stdout=sys.stderr,
+                       check=True)
+
+        treefile = osbuild.ostree.Treefile()
+
+        argv = ["rpm-ostree", "compose", "commit"]
+        argv += [f"--repo={repo}"]
+
+        if parent:
+            argv += [f"--parent={parent}"]
+
+        argv += [f"--write-commitid-to={output_dir}/commit.id",
```

Comment:
Was there the possibility of outputting a tar file instead of a directory? What would be the pros/cons of that? To me seems nice to have a single artefact, but I can imagen there are considerations at play.

Reply:
I would wait until we have assembler pipelines to produce the tar file with an tar assembler stage.